dss/events/handlers/index.py

```diff
@@ -205,6 +201,41 @@ def add_data_to_elasticsearch(bundle_id: str, index_data: dict, index_name: str,
                      json.dumps(index_data, indent=4))
         raise
 
+    try:
+        current_mappings = ElasticsearchClient.get(logger).indices.get_mapping(index_name)[index_name]['mappings']
+        if initial_mappings != current_mappings:
+            refresh_percolate_queries(index_name, replica, logger)
+    except Exception as ex:
+        logger.error("Error refreshing subscription queries for index. Exception: %s, Index name: %s", ex, index_name)
+        raise
+
+
+def refresh_percolate_queries(index_name: str, replica: str, logger) -> None:
+    # When dynamic templates are used and queries for percolation have been added
+    # to an index before the index contains mappings of fields referenced by those queries,
+    # the queries must be reloaded when the mappings are present for the queries to match.
+    # See: https://github.com/elastic/elasticsearch/issues/5750
+    subscription_index_name = Config.get_es_index_name(ESIndexType.subscriptions, Replica[replica])
+    if not ElasticsearchClient.get(logger).indices.exists(subscription_index_name):
+        return
+    subscription_queries = [{'_index': index_name,
+                             '_type': ESDocType.query.name,
+                             '_id': hit['_id'],
+                             '_source': hit['_source']['es_query']
+                             }
+                            for hit in scan(ElasticsearchClient.get(logger),
+                                            index=subscription_index_name,
+                                            doc_type=ESDocType.subscription.name,
+                                            query={'query': {'match_all': {}}})
+                            ]
+
+    if len(subscription_queries) > 0:
+        try:
+            bulk(ElasticsearchClient.get(logger), iter(subscription_queries), refresh=True)
```

Comment:
doing a bulk operation with `refresh=True` sounds really expensive. Does it refresh after each operation or at the end of the bulk?

Reply:
The refresh is performed after the bulk operation, not for each item. This is described in the Elasticsearch documentation here: https://www.elastic.co/guide/en/elasticsearch/reference/2.2/docs-bulk.html#bulk-refresh
This is not expensive.
The refresh is necessary because the next step is to perform the percolate query of the index document against these queries.